src/Ema/Route/Generic/Verification.hs

```diff
@@ -0,0 +1,212 @@
+-}
+type family FieldType (s :: Symbol) (t :: Type) :: Type where
+  FieldType s (GHC.D1 _ (GHC.C1 _ selectors) _) =
+    FieldType s (selectors ())
+  FieldType s ((GHC.S1 ( 'GHC.MetaSel ( 'Just s') _ _ _) (GHC.K1 _ t) GHC.:*: nxt) _) =
+    If (s == s') t (FieldType s (nxt ()))
+  FieldType s (GHC.S1 ( 'GHC.MetaSel ( 'Just s') _ _ _) (GHC.K1 _ t) _) =
+    If (s == s') t (TypeError ("Field selector " P.<> s P.% " does not exist."))
+  FieldType _ _ =
+    TypeError ("Type rep field name lookup: multiple constructors" P.% "")
+
+{- | Traverses the single-constructor generic type representation of a model @r@ to see if at least one of its
+ fields has a sunmodel of type @t@.
+
+ Examples:
+ > data X = X Int | Y deriving GHC.Generic
+ > ContainsSubModel Int (GHC.Rep X ()) == 'False -- X must have only one constructor
+
+ > data X = X {y :: Int, z :: Bool} deriving GHC.Generic
+ > ContainsSubModel Bool (GHC.Rep X ()) == 'True
+ > ContainsSubModel Float (GHC.Rep X ()) == 'False
+-}
+type family ContainsSubModel (t :: Type) (r :: Type) :: Bool where
+  ContainsSubModel () _ =
+    'True
+  ContainsSubModel t (GHC.D1 _ (GHC.C1 _ fields) _) =
+    ContainsSubModel t (fields ())
+  ContainsSubModel t ((GHC.S1 _ (GHC.K1 _ t') GHC.:*: nxt) _) =
+    t == t' || ContainsSubModel t (nxt ())
+  ContainsSubModel t (GHC.S1 _ (GHC.K1 _ t') _) =
+    t == t'
+  ContainsSubModel t _ =
+    'False
+
+type family ModelShapeMismatchError r where
+  ModelShapeMismatchError r = TypeError ("WithSubModelss list does not match the shape of " P.<> r)
+
+{- | @VerifyModels model routeModels lookups@ verifies the given @model@ to ensure that there
+exists a valid @HasSubModels@ instance for the given combination of (model, routeModels, lookups).
+-}
+type family VerifyModels model (routeModels :: [Type]) (lookups :: [Type]) :: Constraint where
+  VerifyModels m '[] '[] = ()
+  VerifyModels m '[] _ = ModelShapeMismatchError m
+  VerifyModels m _ '[] = ModelShapeMismatchError m
+-- TODO: Does not verify if the model is GHC.Generic, and some models don't have /any/ generic instance to begin with,
+-- so if we want to be more robust we can dispatch via class instances. However, the base compiler error should look obvious enough.
+  VerifyModels model (f ': fs) (Proxy (n :: Nat) ': ss) =
+    If
+      (f == Indexed n (GHC.Rep model ()))
+      (VerifyModels model fs ss)
+      ( TypeError
+          ( "Submodel at index " P.<> n P.<> " of " P.<> model
+              P.% "Has type:"
+              P.% Indexed n (GHC.Rep model ())
+              P.% "Subroute specification requires that this instead be:"
+              P.% f
+          )
+      )
+-- TODO: Does not verify if the model is GHC.Generic, and some models don't have /any/ generic instance to begin with,
+-- so if we want to be more robust we can dispatch via class instances. However, the base compiler error should look obvious enough.
+  VerifyModels model (f ': fs) (Proxy (s :: Symbol) ': ss) =
```

Comment:
Is there an example that has the compiler reach this case? I used `Proxy "missing"` but it is throwing `The type Model does not contain a field named 'missing'.` (presumably coming from the HasAny stuff).

Reply:
So this is reaching it correct, but it's actually coughing up 2 errors; the second one is the custom type error. Relevant test case:

```hs
deriveGeneric ''Route
deriveIsRoute ''Route [t|'[WithModel Model, WithSubModels '[(), (), Map Slug Product, Proxy "blah"]]|]
```

Reply:
```hs
src/Ema/Example/Ex03_Store.hs:111:1: error:
    • The type Model does not contain a field named 'blah'.
    • In the third argument of ‘coerce’, namely
<wall of coerce>
      When typechecking the code for ‘subModels’
        in a derived instance for ‘HasSubModels Route’:
        To see the code I am typechecking, use -ddump-deriv
    |
111 | deriveIsRoute
    | ^^^^^^^^^^^^^...

src/Ema/Example/Ex03_Store.hs:111:1: error:
    • Field selector blah does not exist.
    • In the third argument of ‘coerce’, namely
<wall of coerce>
      When typechecking the code for ‘subModels’
        in a derived instance for ‘HasSubModels Route’:
        To see the code I am typechecking, use -ddump-deriv
    |
111 | deriveIsRoute
    | ^^^^^^^^^^^^^...

```

Reply:
Oh, it gets triggered if the field *does* exist, albeit misplaced:

```haskell
     , WithSubModels
        '[ ()
         , ()
         , Map Slug Product
         , Proxy "modelProducts"
         ]
```

```
src/Ema/Example/Ex03_Store.hs:(110,1)-(127,6): error:
    • A field modelProducts of type:
      Map Slug Category
      Does not exist in model 
      Model
```

But the error message is inaccurate. I'll fix that.